### PR TITLE
[Misc] An Example to Compute the Low-noise Perplexity Estimate for Llama-2 model family.

### PR DESCRIPTION
**Pull request by Alexei-V-Ivanov-AMD**
The script examples/measure_pppl_llama2.py produces a realistic perplexity measurement for the quantized KV cache system by processing a sequence of non-overlapping patches of the reference text. Generation of the consecutive symbols in each patch is governed (forced) by the reference text.

The initial context size for the system is set by the parameter "--context-size".

The number of output symbols to generate starting from a given context is set by the parameter "--sample-size". This variable also defines the size of the individual patch. The size of the patch in tokens is equal to the sample size.

For the N-token reference text that is split into M patches with the system's initial context size C, the method takes M*preload + (N-C)*generation time to complete.

Quick correctness validation tips:

Running llama-2-7b-chat-hf model 
( 
    examples/measure_pppl_llama2.py 
    --model=/data/models/llama-2-7b-chat-hf 
    --data=./vllm/tests/prompts/wiki.test.raw 
    --context-size=1024 
    --sample-size=512
)
should result in PPL ~ 6.524227946419175

Running llama-2-7b-chat-hf model 
( 
    examples/measure_pppl_llama2.py 
    --model=/data/models/llama-2-7b-chat-hf 
    --data=./vllm/tests/prompts/wiki.test.raw 
    --context-size=1024 
    --sample-size=512
    --patch-size=1
)
should result in PPL ~ PPL=3.8968611189957523

This testing method is sensitive to the representation precision of the KV cache.
The table below presents perplexities, achieved with different quantization and scaling methods.

<html xmlns:v="urn:schemas-microsoft-com:vml"
xmlns:o="urn:schemas-microsoft-com:office:office"
xmlns:x="urn:schemas-microsoft-com:office:excel"
xmlns="http://www.w3.org/TR/REC-html40">

<head>

<meta name=ProgId content=Excel.Sheet>
<meta name=Generator content="Microsoft Excel 15">
<link id=Main-File rel=Main-File
href="file:///C:/Users/alivanov/AppData/Local/Temp/msohtmlclip1/01/clip.htm">
<link rel=File-List
href="file:///C:/Users/alivanov/AppData/Local/Temp/msohtmlclip1/01/clip_filelist.xml">
<style>

</style>
</head>

<body link="#0563C1" vlink="#954F72" lang=EN-US style='tab-interval:.5in;
word-wrap:break-word'>

llama-2-7b-chat-hf | 647 patches     330849 symb(max)
-- | --
gen 512 X init 1024 | PPLv2
FP8 Scaling 1e3 | 2016.919661
FP8 Scaling 1e2 | 7.110797102
FP8 Scaling 1e1 | 6.550152394
FP16 | 6.524227946
FP8 | 6.541197624
FP8 Scaling 1e-1 | 6.545720813
FP8 Scaling 1e-2 | 57.70005660

</body>

</html>

FILL IN THE PR DESCRIPTION HERE

FIX #xxxx (*link existing issues this PR will resolve*)

**BEFORE SUBMITTING, PLEASE READ THE CHECKLIST BELOW AND FILL IN THE DESCRIPTION ABOVE**

---

<details>

<summary><b> PR Checklist (Click to Expand) </b></summary>

<p>Thank you for your contribution to vLLM! Before submitting the pull request, please ensure the PR meets the following criteria. This helps vLLM maintain the code quality and improve the efficiency of the review process.</p>

<h3>PR Title and Classification</h3>
<p>Only specific types of PRs will be reviewed. The PR title is prefixed appropriately to indicate the type of change. Please use one of the following:</p>
<ul>
    <li><code>[Bugfix]</code> for bug fixes.</li>
    <li><code>[CI/Build]</code> for build or continuous integration improvements.</li>
    <li><code>[Doc]</code> for documentation fixes and improvements.</li>
    <li><code>[Model]</code> for adding a new model or improving an existing model. Model name should appear in the title.</li>
    <li><code>[Frontend]</code> For changes on the vLLM frontend (e.g., OpenAI API server, <code>LLM</code> class, etc.) </li>
    <li><code>[Kernel]</code> for changes affecting CUDA kernels or other compute kernels.</li>
    <li><code>[Core]</code> for changes in the core vLLM logic (e.g., <code>LLMEngine</code>, <code>AsyncLLMEngine</code>, <code>Scheduler</code>, etc.)</li>
    <li><code>[Hardware][Vendor]</code> for hardware-specific changes. Vendor name should appear in the prefix (e.g., <code>[Hardware][AMD]</code>).</li>
    <li><code>[Misc]</code> for PRs that do not fit the above categories. Please use this sparingly.</li>
</ul>
<p><strong>Note:</strong> If the PR spans more than one category, please include all relevant prefixes.</p>

<h3>Code Quality</h3>

<p>The PR need to meet the following code quality standards:</p>

<ul>
    <li>We adhere to <a href="https://google.github.io/styleguide/pyguide.html">Google Python style guide</a> and <a href="https://google.github.io/styleguide/cppguide.html">Google C++ style guide</a>.</li>
    <li>Pass all linter checks. Please use <a href="https://github.com/vllm-project/vllm/blob/main/format.sh"><code>format.sh</code></a> to format your code.</li>
    <li>The code need to be well-documented to ensure future contributors can easily understand the code.</li>
    <li>Include sufficient tests to ensure the project to stay correct and robust. This includes both unit tests and integration tests.</li>
    <li>Please add documentation to <code>docs/source/</code> if the PR modifies the user-facing behaviors of vLLM. It helps vLLM user understand and utilize the new features or changes.</li>
</ul>

<h3>Notes for Large Changes</h3>
<p>Please keep the changes as concise as possible. For major architectural changes (>500 LOC excluding kernel/data/config/test), we would expect a GitHub issue (RFC) discussing the technical design and justification. Otherwise, we will tag it with <code>rfc-required</code> and might not go through the PR.</p>

<h3>What to Expect for the Reviews</h3>

<p>The goal of the vLLM team is to be a <i>transparent reviewing machine</i>. We would like to make the review process transparent and efficient and make sure no contributor feel confused or frustrated. However, the vLLM team is small, so we need to prioritize some PRs over others. Here is what you can expect from the review process: </p>

<ul>
    <li> After the PR is submitted, the PR will be assigned to a reviewer. Every reviewer will pick up the PRs based on their expertise and availability.</li>
    <li> After the PR is assigned, the reviewer will provide status update every 2-3 days. If the PR is not reviewed within 7 days, please feel free to ping the reviewer or the vLLM team.</li>
    <li> After the review, the reviewer will put an <code> action-required</code> label on the PR if there are changes required. The contributor should address the comments and ping the reviewer to re-review the PR.</li>
    <li> Please respond to all comments within a reasonable time frame. If a comment isn't clear or you disagree with a suggestion, feel free to ask for clarification or discuss the suggestion.
 </li>
</ul>

<h3>Thank You</h3>

<p> Finally, thank you for taking the time to read these guidelines and for your interest in contributing to vLLM. Your contributions make vLLM a great tool for everyone! </p>

</details>

### CODE REVIEW

**Comment by Alexei-V-Ivanov-AMD** [comment]
Applying all comments and suggestions from the previous iteration 
(https://github.com/vllm-project/vllm/pull/3730)

**Comment by robertgshaw2-neuralmagic** [comment]
Is there a reason why we cannot use `lm-eval` to run this:

Install:
```bash
pip install vllm lm_eval
```

Launch:
```bash
meta-llama/Meta-Llama-3.1-8B-Instruct
lm_eval --model vllm --tasks wikitext --model_args pretrained=$MODEL,enable_chunked_prefill=true
```

Output:

```bash
vllm (pretrained=meta-llama/Meta-Llama-3.1-8B-Instruct,enable_chunked_prefill=true), gen_kwargs: (None), limit: None, num_fewshot: None, batch_size: 1
| Tasks  |Version|Filter|n-shot|    Metric     |   |Value |   |Stderr|
|--------|------:|------|-----:|---------------|---|-----:|---|------|
|wikitext|      2|none  |     0|bits_per_byte  |↓  |0.5818|±  |   N/A|
|        |       |none  |     0|byte_perplexity|↓  |1.4967|±  |   N/A|
|        |       |none  |     0|word_perplexity|↓  |8.6413|±  |   N/A|
```

**Comment by simon-mo** [comment]
Ah i didn't know lm eval can do exactly this. lm eval is well maintained and standardized. I think we should close this PR (sorry!),.

**Comment by robertgshaw2-neuralmagic** [comment]
Actually, now that I look at it, this is PPPL as compared to PPL

**Comment by robertgshaw2-neuralmagic** [comment]
@Alexei-V-Ivanov-AMD what is the advantage of PPPL as compared to PPL?

**Comment by simon-mo** [comment]
re-opened for discussion

**Comment by Alexei-V-Ivanov-AMD** [comment]
> @Alexei-V-Ivanov-AMD what is the advantage of PPPL as compared to PPL?

@robertgshaw2-neuralmagic the description of the PR clearly states it: 
"a realistic perplexity measurement for the quantized KV cache system by processing a sequence of non-overlapping patches of the reference text. Generation of the consecutive symbols in each patch is governed (forced) by the reference text."

In the essence, with PPPL (and not PPL) one can evaluate the performance of the engine in the sustained autoregressive generation mode, not only pre-loading and generating the first output.

**Comment by robertgshaw2-neuralmagic** [comment]
> governed

Couldn't we accomplish the same by running with chunked prefill?

**Comment by robertgshaw2-neuralmagic** [comment]
> > @Alexei-V-Ivanov-AMD what is the advantage of PPPL as compared to PPL?
> 
> @robertgshaw2-neuralmagic the description of the PR clearly states it: "a realistic perplexity measurement for the quantized KV cache system by processing a sequence of non-overlapping patches of the reference text. Generation of the consecutive symbols in each patch is governed (forced) by the reference text."
> 
> In the essence, with PPPL (and not PPL) one can evaluate the performance of the engine in the sustained autoregressive generation mode, not only pre-loading and generating the first output.

Can't we accomplish this exact same thing by running with `chunked_prefill`?

**Comment by Alexei-V-Ivanov-AMD** [comment]
>Can't we accomplish this exact same thing by running with --enable-chunked-prefill=True?

No. Chunked pre-fill is not capable of doing the "exact same thing". Can you describe the mechanism with which you believe the chunked pre-fill can accomplish the same result?

**Comment by robertgshaw2-neuralmagic** [comment]
> > Can't we accomplish this exact same thing by running with --enable-chunked-prefill=True?
> 
> No. Chunked pre-fill is not capable of doing the "exact same thing". Can you describe the mechanism with which you believe the chunked pre-fill can accomplish the same result?

With chunked prefill, we run prefill for N tokens at a time. As a result, whenever we are computing token index `i=N*k+j` (that is, the `jth` token of the `kth` chunk), all the tokens at indices `[0,N*k]` will be in KV cache. If we set the chunk_size to 1 (i.e. `N=1`), I think we are computing `PPPL` as you described.

**Comment by Alexei-V-Ivanov-AMD** [comment]
>If we set the chunk_size to 1 (i.e. N=1). I think we are computing PPPL as you describe

That statement needs empirical confirmation (i.e. the extensive experimentation). If I understood your explanation correctly, then (IMHO) the PPPL could give close, but not identical (!) results.

However, there is a standard list of benefits in measuring the actual sustained autoregressive operation:

1. Efficiency. To measure perplexity in autoregressive operation we just turn PPPL on, i.e. only slightly altering the sampling and, thus, making the measurement and aggregation on the actual auto-regressive system. The PPPL provides the minimal overhead compared to the actual operation, thus, it is maximizing *Efficiency*. 

2. Validity. We don't need to prove the equivalence theorem for the "chunked prefill" emulator. Without such proof the *validity* of the test is always a question.

3. Accuracy. Because of the approximate nature of the floating point and mixed precision calculation one can't hope for the uniform equivalence of the computation sequences even if those are supposed to be equivalent arithmetically (assuming infinite-precision representation). This aspect complicates (if not makes outright impossible) proving the equivalence theorem that I'm talking above. Instead of hoping to get a close overall outcome while analyzing the approximate FP and MP computation schedules, one has to concentrate on the detection and analysis of the sources of noise that is injected through the real computational pipeline. Our empirical results show that the PPPL has less internal (test-specific) noise, which allows it to systematically attain better perplexity figures w.r.t. the classic PPL. That makes this method a better tool to analyze the said sources of noise.

**Comment by simon-mo** [comment]
As explained a bit offline, I think the force sampling method will likely to be removed due simplicity and maintainability reason. I believe you can implement exactly the same functionality using logits processor API

https://github.com/vllm-project/vllm/blob/baa5467547a758af35f442af6edfbc0fb73c83ce/vllm/model_executor/layers/logits_processor.py#L137-L143

This is where you can pass in a sampling params with a custom callback function, the function can take two or three arguments, and mutate the logits to return exactly the output needed. I believe that achieves something in this effect?
